libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotter.kt

```diff
@@ -5,112 +5,170 @@
+    }
 
-        val DEFAULT_CLASS_FILTER = { invariantSeparatorsRelativePath: String, isDirectory: Boolean ->
-            !isDirectory
-                    && invariantSeparatorsRelativePath.endsWith(".class", ignoreCase = true)
-                    && !invariantSeparatorsRelativePath.endsWith("module-info.class", ignoreCase = true)
-                    && !invariantSeparatorsRelativePath.startsWith("meta-inf", ignoreCase = true)
+    /** Creates [KotlinClassSnapshot] of the given class, or returns `null` if the class is not a Kotlin class. */
+    private fun trySnapshotKotlinClass(clazz: ClassFileWithContents): KotlinClassSnapshot? {
+        return KotlinClassInfo.tryCreateFrom(clazz.contents)?.let {
+            KotlinClassSnapshot(it)
         }
+    }
 
-        /** Creates a [ClasspathEntryContentsReader] for the given classpath entry (directory or jar). */
-        fun from(classpathEntry: File): ClasspathEntryContentsReader {
-            return if (classpathEntry.isDirectory) {
-                DirectoryContentsReader(classpathEntry)
+    /**
+     * Creates [JavaClassSnapshot]s of the given Java classes.
+     *
+     * Note that creating a [JavaClassSnapshot] for a nested class will require accessing the outer class (and possibly vice versa).
+     * Therefore, outer classes and nested classes must be passed together in one invocation of this method.
+     */
+    private fun snapshotJavaClasses(classes: List<ClassFileWithContents>): List<JavaClassSnapshot> {
+        val classFiles = classes.map { it.classFile }
+        val classesContents = classes.map { it.contents }
+        val classNames = classesContents.map { JavaClassName.compute(it) }
+        val classIds = computeJavaClassIds(classNames)
+
+        // Snapshot special cases first (map a class index to its snapshot, or `null` if it will be created in the next round)
+        val specialCaseSnapshots: Map<Int, JavaClassSnapshot?> = classFiles.indices.associateWith { index ->
+            val className = classNames[index]
+            val classId = classIds[index]
+            if (classId.isLocal) {
+                // A local class can't be compiled against, so any changes in a local class will not cause recompilation of other classes.
+                // Note that in this context, a nested class of a local class is also considered local (see ClassId's kdoc). Therefore, we
+                // checked `classId.isLocal` instead of `className is LocalClass`.
+                EmptyJavaClassSnapshot
             } else {
-                JarContentsReader(classpathEntry)
+                when (className) {
+                    is TopLevelClass -> null
+                    is NestedNonLocalClass -> {
+                        if (className.isPossiblyAnonymous) {
+                            // TODO: Currently BinaryJavaClass is not able to handle anonymous classes, fix this later.
```

Comment:
We can ignore anonymous classes, they cannot impact compilation.

Reply:
Ah yes, my only remaining concern is that we can't tell if a class is anonymous precisely. I've changed this to `className.simpleName.toIntOrNull() != null` and added a comment:

"A nested class whose simple name is a number also can't be referenced from other source files (whether it is an anonymous class or not). Therefore, the snapshot of this type of class is also empty."

Reply:
You can tell if a class is anonymous from `visitInnerClass(name: String, outerName: String?, innerName: String?, access: Int)` because `innerName` will be null - https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.6.

Reply:
Yes, I used `innerName` at first, but `groovy-all-2.4.8.jar` has `org.codehaus.groovy.transform.tailrec.VariableReplacedListener$1` which is an anonymous class but `innerName == "1"` for some reason (`IncrementalCompilationClasspathSnapshotJvmMultiProjectIT.testCompileLibWithGroovy` would fail if it was not handled): 
```
interface VariableReplacedListener {
    void variableReplaced(VariableExpression oldVar, VariableExpression newVar)

    static VariableReplacedListener NULL = new VariableReplacedListener() {
        @Override
        void variableReplaced(VariableExpression oldVar, VariableExpression newVar) {
            //do nothing
        }
    }
}
```

Other anonymous classes in that jar still have `innerName == null`.

I just updated the check to `val isAnonymous = (innerName == null) || (innerName.toIntOrNull() != null)` and added a comment.

Reply:
What about `outerName`? The spec says it should be null for local/anonymous classes -> `If C is not a member of a class or an interface (that is, if C is a top-level class or interface (JLS §7.6) or a local class (JLS §14.3) or an anonymous class (JLS §15.9.5)), the value of the outer_class_info_index item must be zero.`.

Reply:
The above class has `outerName != null`. I just doube-checked and got:

```
name = "org/codehaus/groovy/transform/tailrec/VariableReplacedListener$1"
outerName = "org/codehaus/groovy/transform/tailrec/VariableReplacedListener"
innerName = "1"
access = 1
```

Looks like it doesn't meet the specs, but we'll probably still have to handle it.

Reply:
As discussed offline, I've updated this PR to ignore anonymous & synthetic classes as they won't impact recompilation. 

Side note: `org/codehaus/groovy/transform/tailrec/VariableReplacedListener$1` is non-synthetic, but `org/codehaus/groovy/transform/tailrec/VariableReplacedListener$2` is synthetic, and `BinaryJavaClass` currently ignores synthetic classes.